components/batch-system/src/batch.rs

```diff
@@ -288,21 +306,31 @@ impl<N: Fsm, C: Fsm, Handler: PollHandler<N, C>> Poller<N, C, Handler> {
             if !batch.normals.is_empty() {
                 for (i, p) in batch.normals.iter_mut().enumerate() {
                     let len = self.handler.handle_normal(p);
```

Comment:
Could we record len and counter in Fsm? I think this would be more clear..
```
pub trait Fsm {
      fn get_reset_messages() -> usize;
      fn get_counter() -> usize;
}; 

Impl Fsm for ApplyFsm {
     fn begin(..) {
          self.rest_messages = None;
     }
     fn handle_normal(...) {
          self.counter += 1;
          self.rest_messages = Some(0);
     }
     fn release() {
          self.counter = 0;
     }
}
```

Reply:
`len` will be removed in the future by using atomic state. I don't think `Fsm` should care about counter, it's a scheduling metrics of batch system.

Reply:
Ok. How about  replacing two vec with `Vec<(Box<Fsm>, usize)> `

Reply:
The fsm vec is used by handler in the `end` hook. Combining them together will require copy for every invocation to `end`.